diff --git a/build/Android.common_build.mk b/build/Android.common_build.mk
index d82aab5..fea1666 100644
--- a/build/Android.common_build.mk
+++ b/build/Android.common_build.mk
@@ -116,7 +116,7 @@ endif
 
 # Host.
 ART_HOST_CLANG := false
-ifneq ($(WITHOUT_HOST_CLANG),true)
+ifeq ($(WITHOUT_HOST_CLANG),false)
   # By default, host builds use clang for better warnings.
   ART_HOST_CLANG := true
 endif
